### PR TITLE
Use spans pointing at the inside of a rustdoc attribute

### PR DESCRIPTION
**Pull request by estebank**
Follow up to #51111.

Point to the link in a rustdoc attribute where intralink resolution failed, instead of the full rustdoc attribute's span.

r? @GuillaumeGomez cc @kennytm

### CODE REVIEW

**Comment by estebank** [comment]
The code assumes that the leading whitespace is the same throughout the doc comment. The code could be tweaked to correctly handle the cases where this is not the case, but I'm going to bed now :)

**Comment by GuillaumeGomez** [comment]
Awesome! Once ci fixed, r=me

**Comment by kennytm** [comment]
@estebank Could you verify how this works for all kinds of doc comments:

```rust
#![feature(external_doc)]

/// Foo
/// bar [BarA] bar
/// baz
pub fn a() {}

/**
 * Foo
 * bar [BarB] bar
 * baz
 */
pub fn b() {}

/** Foo

bar [BarC] bar
baz

    let bar_c_1 = 0;
    let bar_c_2 = 0;
    let g = [bar_c_1];
    let h = g[bar_c_2];

*/
pub fn c() {}

#[doc = "Foo\nbar [BarD] bar\nbaz"]
pub fn d() {}

#[doc(include = "file.md")]
pub fn e() {}

macro_rules! f {
    ($f:expr) => {
        #[doc = $f]
        pub fn f() {}
    }
}
f!("Foo\nbar [BarF] bar\nbaz");
```

Here the content of `file.md` is 

```markdown
Foo
bar [BarE] bar
baz
```

<details><summary>For reference, the existing output post-51111</summary>

```
warning: [BarA] cannot be resolved, ignoring it...
 --> 1.rs:3:1
  |
3 | / /// Foo
4 | | /// bar [BarA] bar
5 | | /// baz
  | |_______^
  |
  = note: the link appears in this line:
          
           bar [BarA] bar
                ^^^^

warning: [BarB] cannot be resolved, ignoring it...
  --> 1.rs:8:1
   |
8  | / /**
9  | |  * Foo
10 | |  * bar [BarB] bar
11 | |  * baz
12 | |  */
   | |___^
   |
   = note: the link appears in this line:
           
            bar [BarB] bar
                 ^^^^

warning: [BarC] cannot be resolved, ignoring it...
  --> 1.rs:15:1
   |
15 | / /** Foo
16 | |
17 | | bar [BarC] bar
18 | | baz
...  |
24 | |
25 | | */
   | |__^
   |
   = note: the link appears in this line:
           
           bar [BarC] bar
                ^^^^

warning: [BarD] cannot be resolved, ignoring it...
  --> 1.rs:28:1
   |
28 | #[doc = "Foo\nbar [BarD] bar\nbaz"]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: the link appears in this line:
           
           bar [BarD] bar
                ^^^^

warning: [BarE] cannot be resolved, ignoring it...
  --> 1.rs:31:1
   |
31 | #[doc(include = "file.md")]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: the link appears in this line:
           
           bar [BarE] bar
                ^^^^

warning: [BarF] cannot be resolved, ignoring it...
  --> 1.rs:36:9
   |
36 |         #[doc = $f]
   |         ^^^^^^^^^^^
...
40 | f!("Foo\nbar [BarF] bar\nbaz");
   | ------------------------------- in this macro invocation
   |
   = note: the link appears in this line:
           
           bar [BarF] bar
                ^^^^
```

</details>

**Comment by kennytm** [comment]
@estebank Could you verify how this works for all kinds of doc comments:

```rust
#![feature(external_doc)]

/// Foo
/// bar [BarA] bar
/// baz
pub fn a() {}

/**
 * Foo
 * bar [BarB] bar
 * baz
 */
pub fn b() {}

/** Foo

bar [BarC] bar
baz

    let bar_c_1 = 0;
    let bar_c_2 = 0;
    let g = [bar_c_1];
    let h = g[bar_c_2];

*/
pub fn c() {}

#[doc = "Foo\nbar [BarD] bar\nbaz"]
pub fn d() {}

#[doc(include = "file.md")]
pub fn e() {}

macro_rules! f {
    ($f:expr) => {
        #[doc = $f]
        pub fn f() {}
    }
}
f!("Foo\nbar [BarF] bar\nbaz");
```

Here the content of `file.md` is 

```markdown
Foo
bar [BarE] bar
baz
```

<details><summary>For reference, the existing output post-51111</summary>

```
warning: [BarA] cannot be resolved, ignoring it...
 --> 1.rs:3:1
  |
3 | / /// Foo
4 | | /// bar [BarA] bar
5 | | /// baz
  | |_______^
  |
  = note: the link appears in this line:
          
           bar [BarA] bar
                ^^^^

warning: [BarB] cannot be resolved, ignoring it...
  --> 1.rs:8:1
   |
8  | / /**
9  | |  * Foo
10 | |  * bar [BarB] bar
11 | |  * baz
12 | |  */
   | |___^
   |
   = note: the link appears in this line:
           
            bar [BarB] bar
                 ^^^^

warning: [BarC] cannot be resolved, ignoring it...
  --> 1.rs:15:1
   |
15 | / /** Foo
16 | |
17 | | bar [BarC] bar
18 | | baz
...  |
24 | |
25 | | */
   | |__^
   |
   = note: the link appears in this line:
           
           bar [BarC] bar
                ^^^^

warning: [BarD] cannot be resolved, ignoring it...
  --> 1.rs:28:1
   |
28 | #[doc = "Foo\nbar [BarD] bar\nbaz"]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: the link appears in this line:
           
           bar [BarD] bar
                ^^^^

warning: [BarE] cannot be resolved, ignoring it...
  --> 1.rs:31:1
   |
31 | #[doc(include = "file.md")]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: the link appears in this line:
           
           bar [BarE] bar
                ^^^^

warning: [BarF] cannot be resolved, ignoring it...
  --> 1.rs:36:9
   |
36 |         #[doc = $f]
   |         ^^^^^^^^^^^
...
40 | f!("Foo\nbar [BarF] bar\nbaz");
   | ------------------------------- in this macro invocation
   |
   = note: the link appears in this line:
           
           bar [BarF] bar
                ^^^^
```

</details>

File: src/librustdoc/clean/mod.rs [hunk 1]
```diff
@@ -1191,30 +1191,59 @@ fn resolution_failure(
...
+                    code_dox.lines().skip(1).take(line_offset - 1).fold(0, |sum, line| {
+                        sum + doc_comment_padding + line.len() - line.trim().len()
+                    })
+            };
 
-    }
+            // Extract the specific span
+            let lo = sp.lo() + syntax_pos::BytePos((link_range.start + code_dox_len) as u32);
+            let hi = lo + syntax_pos::BytePos(link_range.len() as u32);
+            let sp = sp.with_lo(lo).with_hi(hi);
```
Comment (kennytm) [inline src/librustdoc/clean/mod.rs]: You could use 

```rust
Span::from_inner_byte_pos(
    link_range.start + code_dox_len, 
    link_range.end + code_dox_len,
)
```

here 😃 

https://doc.rust-lang.org/nightly/nightly-rustc/syntax_pos/struct.Span.html#method.from_inner_byte_pos

**Review by kennytm — APPROVED** [review]

**Comment by kennytm** [comment]
@bors retry 

Why closing the tree would cancel the existing PR? 😕

**Pull request merged**